vscode-extension/src/extension.ts

```diff
@@ -1,27 +1,162 @@
-// Import the module and reference it with the alias vscode in your code below
-import * as vscode from 'vscode';
+import * as util from 'util';
+import * as cp from 'child_process';
+import { workspace, window, commands, Uri, ExtensionContext, QuickPickItem } from 'vscode';
+import { setupDevContainerFiles, readDevboxJson } from './devcontainer';
 
 // This method is called when your extension is activated
 // Your extension is activated the very first time the command is executed
-export function activate(context: vscode.ExtensionContext) {
-
-	// Use the console to output diagnostic information (console.log) and errors (console.error)
+export function activate(context: ExtensionContext) {
 	// This line of code will only be executed once when your extension is activated
-	vscode.window.onDidOpenTerminal(async (event) => {
-		runDevboxShell();
+	initialCheckDevboxJSON();
+	// Creating file watchers to watch for events on devbox.json
+	const fswatcher = workspace.createFileSystemWatcher("**/devbox.json", false, false, false);
+	fswatcher.onDidDelete(e => commands.executeCommand('setContext', 'devbox.configFileExists', false));
+	fswatcher.onDidCreate(e => commands.executeCommand('setContext', 'devbox.configFileExists', true));
+	fswatcher.onDidChange(e => initialCheckDevboxJSON());
+
+	// Check for devbox.json when a new folder is opened
+	workspace.onDidChangeWorkspaceFolders(async (e) => initialCheckDevboxJSON());
+
+	// run devbox shell when terminal is opened
+	window.onDidOpenTerminal(async (e) => {
+		if (workspace.getConfiguration("devbox").get("autoShellOnTerminal")) {
+			await runInTerminal('devbox shell');
+		}
+	});
+
+	const devboxAdd = commands.registerCommand('devbox.add', async () => {
+		const result = await window.showInputBox({
+			value: '',
+			placeHolder: 'Package to add to devbox. E.g., python39',
+		});
+		await runInTerminal(`devbox add ${result}`);
+	});
+
+	const devboxRun = commands.registerCommand('devbox.run', async () => {
+		const items = await getDevboxScripts();
+		if (items.length > 0) {
+			const result = await window.showQuickPick(items);
+			await runInTerminal(`devbox run ${result}`);
+		} else {
+			window.showInformationMessage("No scripts found in devbox.json");
+		}
 	});
+
+	const devboxShell = commands.registerCommand('devbox.shell', async () => {
+		// todo: add support for --config path to devbox.json
+		await runInTerminal('devbox shell');
+	});
+
+	const devboxRemove = commands.registerCommand('devbox.remove', async () => {
+		const items = await getDevboxPackages();
+		if (items.length > 0) {
+			const result = await window.showQuickPick(items);
+			await runInTerminal(`devbox rm ${result}`);
+		} else {
+			window.showInformationMessage("No packages found in devbox.json");
+		}
+	});
+
+	const devboxInstall = commands.registerCommand('devbox.install', async () => {
+		// todo: add support for --config path to devbox.json
+		await runInTerminal('devbox shell -- echo "Packages installed""');
```

Comment:
Not sure if we should have `devbox install` command at all. It doesn't do what a user might expect it to do, since `devbox shell -- echo "Packages installed"` installs the nix packages but does not expose them in the current PATH. So from a user's perspective, nothing is installed. Thoughts?

Reply:
yeah that's fair I'll remove it 👍